### ISSUE TITLE
Data styling – labeling problem with closed ways

### ISSUE THREAD

**Issue by jajajaneeneenee** [MarcusWolschon/osmeditor4android#1397]
## Vespucci Version
15.2.1

## Behaviour/Symptoms
If you want to set a labelKey attribute for a closed way in a data styling xml file, and it's **not** `labelKey="name"`, the label will not be displayed if the closed way doesn't have a name, too. Other labelKeys than "name" only seem to work fine on nodes (it also works if a node has no name). But ways (and only closed ways seem to work at all, but I'm not sure) seem to have a problem ...

Example:
I wanted to change the standard styling for a parking area
`<feature type="way" tags="amenity=parking" color="ff99CCFF"  />`
to this:
`<feature type="way" tags="amenity=parking" closed="true" color="99709fce" style="FILL_AND_STROKE" iconPath="preset" labelKey="operator" />`

And everything works fine (including the icon) except the labelKey: if a parking area has an operator attribute, it is only displayed if the parking ALSO has a name attribute. Otherwise only the icon will be shown without any label.

And with `labelKey="name"` it works fine:
`<feature type="way" tags="amenity=parking" closed="true" color="99709fce" style="FILL_AND_STROKE" iconPath="preset" labelKey="name" />`

**Comment by simonpoole** [comment]
This took some staring at the code to realize what the issue is, but I assume you not actually getting any of the rest of the styling either? 

Because as far as I can tell the reason is simply because on line 128 there is a default entry for closed amenity ways which sets the labelKey to "name", and closed amenity=parking ways are matching with that. Your entry should hierarchically be nested in that and not where the current "general" amenity=parking is.

It is a bug in so far as that there should be specific styling for the typical closed way amenites in the style files.

**Comment by simonpoole** [comment]
Rendering added in https://github.com/MarcusWolschon/osmeditor4android/commit/4922b782fc86d420293fb6504522bb8977ad1fba, inheritance bug fixed in https://github.com/MarcusWolschon/osmeditor4android/commit/379df85ea13d5eaad1ffb5f6398fed2b8a89aa5a

**Comment by jajajaneeneenee** [comment]
Hello!

Thank you for fixing the inheritance bug – I hope it was really the source of the problem and/or its complete solution ... but I guess you checked it out ...

First I have to admin that I do not understand completely what you mean with "but I assume you not actually getting any of the rest of the styling either?". (I mean – I should have posted the complete code if the sytling xml here? Or do you ask if I took a look at the complete code?) Google translator also didn't help in this case ... sometimes my english skills are perhaps not good enough.

But I understand what you mean with the code in line 128 and the need of hierachially nesting the code.

Only to make it clear:
- I'm usually quite painstaking when checking things about a problem (and try out every combination which perhaps could be involved in the problem – as far as I can assume of course).
- I also saw code line 128 (with labelKey="name" for amenities) and I deleted it (commented it out) and replaced it with this nested code lines (this is my complete code for amenities – I liked to have it with FILL_AND_STROKE and an icon also ...):

```
	<feature type="way" tags="amenity" updateWidth="true" widthFactor="0.8" color="ffCCCC00" style="STROKE" cap="BUTT" join="MITER" iconPath="preset" labelKey="name">
		<dash phase="1.0">
			<interval length="2.0" />
			<interval length="2.0" />
		</dash>
		<feature type="way" tags="amenity=parking" closed="true" color="99709fce" casingStyle="parking_casing" style="FILL_AND_STROKE" iconPath="preset" labelKey="operator" />
		<feature type="way" tags="amenity=bicycle_parking" closed="true" color="9999CCFF" casingStyle="parking_casing" style="FILL_AND_STROKE" iconPath="preset" labelKey="operator"/>
		<feature type="way" tags="amenity=motorcycle_parking" closed="true" color="9999CCFF" casingStyle="parking_casing" style="FILL_AND_STROKE" iconPath="preset" labelKey="operator" />
		<feature type="way" tags="amenity=grave_yard" updateWidth="true" widthFactor="0.5" color="ff718a59" style="STROKE" cap="BUTT" join="MITER" iconPath="preset" labelKey="name" phase="">
			<feature type="way" tags="amenity=grave_yard" area="true" closed="true" color="88718a59" style="FILL" casingStyle="grave_yard_casing" iconPath="preset" labelKey="name" />
		</feature>
	</feature>
```

- **And** I also made a test **without** `labelKey="name"` in the parent feature code line (the first one).
- **And** I also tried it **ONLY** with the line (I mean no other amenity code line at all!): 
`<feature type="way" tags="amenity=parking" closed="true" color="99709fce" casingStyle="parking_casing" style="FILL_AND_STROKE" iconPath="preset" labelKey="operator" />`

- I even deleted this line to be sure that it does not have any influence:
`<feature type="way" tags="building|amenity" closed="true" iconPath="preset" labelKey="name"/>`
- And I found no other code in the styling xml which perhaps could have an influence here (but perhaps I am missing something).

**The result was:
In ALL cases, the operator text is only shown if the parking also has a name! I re-checked it now (with Vespucci 15.2.2) and it's really like this. I hope I did not overlook something ....**

And after all these tests, I decided to post it here as an issue ... And as I already wrote: with nodes it's no problem at all to use any key for labelling – despite the line 113: ` <feature type="node" tags="name" labelKey="name" />` (and no nesting of other labelling code lines for nodes).
